### ISSUE TITLE
Significant digits being lost in transformations

### ISSUE THREAD

**Issue by jcchavezs** [hypertrace/hypertrace#93]
## Bug

There is a discrepancy over reported data and transformed data between jaeger format and hypertrace. Look at this trace from jaeger which shows `343249` as duration (in microseconds)

```json
{
    "traceID": "65272b59f5737e8d6b6e382622ab5735",
    "spanID": "52dc390b398f4b0f",
    "operationName": "Sent.hipstershop.CheckoutService.PlaceOrder",
    "references": [
        {
            "refType": "CHILD_OF",
            "traceID": "65272b59f5737e8d6b6e382622ab5735",
            "spanID": "70c356eeecde9395"
        }
    ],
    "startTime": 1602566264242895,
    "duration": 343249,
    "tags": [
        {
            "key": "Client",
            "type": "bool",
            "value": true
        },
        {
            "key": "FailFast",
            "type": "bool",
            "value": true
        },
        {
            "key": "status.code",
            "type": "int64",
            "value": 0
        },
        {
            "key": "status.message",
            "type": "string",
            "value": ""
        },
        {
            "key": "internal.span.format",
            "type": "string",
            "value": "jaeger"
        }
    ],
    "logs": [],
    "processID": "p8",
    "warnings": null
},
```

where as what hypertrace retrieves from the graphql query is `344`:

```json
{
    "id": "52dc390b398f4b0f",
    "displaySpanName": "hipstershop.CheckoutService.PlaceOrder",
    "duration": 344,
    "endTime": 1602566264586,
    "parentSpanId": "70c356eeecde9395",
    "serviceName": "traceshop",
    "spanTags": {
        "jaeger.servicename": "frontend",
        "failfast": "true",
        "client": "true",
        "status.message": "",
        "status.code": "0",
        "span.status": ""
    },
    "startTime": 1602566264242,
    "type": "EXIT",
    "traceId": "65272b59f5737e8d6b6e382622ab5735",
    "__typename": "Span"
},
```

Notice `endTime` is also different: `1602566264242895` (jaeger) `1602566264242` (hypertrace).

## Proposal
I believe the reason is that we hypertrace, at some point in the pipeline or the graphql API is trying to convert data into miliseconds in a weird way. Ideally we should keep the data in the units origin reports and also follow what most of the tracers do which is `microseconds` (expect for example `datadog`).

## Questions to address (if any)
- Where does this transformation happens?

cc @aaron-steinfeld @kotharironak @adriancole @laxmanchekka @jake-bassett

**Comment by aaron-steinfeld** [comment]
Holding more sig figs does make sense, but I think keeping maintaining an unknown origin unit may be difficult in terms of standardization. I'd be in favor of storing in the highest typical resolution - nanos?

I was curious on this one, so started digging around a bit, it looks like the collector converts it to the jaeger proto format (nano res):
https://github.com/hypertrace/opentelemetry-collector/blob/main/translator/trace/jaeger/traces_to_jaegerproto.go#L200

which is later converted to the internal format (ms res):
I believe this is where the conversion happens: https://github.com/hypertrace/span-normalizer/blob/main/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java#L192

Using the standard conversion utility associated with that data type:
https://github.com/protocolbuffers/protobuf/blob/master/java/util/src/main/java/com/google/protobuf/util/Timestamps.java#L316

The behavior we see does match the docs on that util, but may not be what is expected. All of it happens very early on in the ingest pipeline.

**Comment by buchi-busireddy** [comment]
Good catch @jcchavezs and your analysis is correct @aaron-steinfeld 
It seems like we should modify the data model for HT platform internal span to maintain at lower res (nanos like Aaron said) and migrate all jobs and services to use this field. Will be a good story for anyone who wants to understand the entire Hypertrace end-to-end.

**Comment by jcchavezs** [comment]
I honestly don't think nano is a good choice, I don't see what is the win
of supporting it since most instrumentation report microtime (so we might
have to multiply by 1000 every time).

In one hand, nano values are meaningless from user perspective (and
probably for automated analysis too) and also language support is at least
questionable. Also, zipkin, Jaeger, haystack and others support micro and
if that is not a standard, I am not sure what it is.

In sum, supporting nanos not only is meaningless for the user but also
cumbersome from technical perspective.

References:

- https://openbase.io/js/nano-time/documentation
- https://vstinner.github.io/python37-pep-564-nanoseconds.html
- https://thephp.cc/news/2020/06/high-resolution-monotonic-timer

**Comment by adriancole** [comment]
@jcchavezs I agree epoch nanos is kind of silly. We are also mixing timestamps and durations.. Let's focus on the timestamp one.

Zipkin timestamp is epoch micros. This serves two purposes:
* Represents the highest usual accuracy of a system clock
* Using the maximum safe integer in JSON (2^53 - 1), epoch micros lasts until 2255-06-05 UTC (if I did math right!)

Nanos doesn't cover either of these. It introduces conversion problems in timestamps. It can make sense for duration, though, but hardly worth mixing units.

Below are a dump of notes. TL;DR; is I think everything in boils down to some folks inheriting google's proto timestamp, which allows nanos offset for epoch seconds, then trying to work around the performance impacts of that choice.

---

I started questioning this [a long time ago](https://issues.apache.org/jira/browse/HTRACE-376?focusedCommentId=15356360&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-15356360) in the now defunct HTRACE project. I think it boils down to culture. Does the culture perceive it will get accuracy to nanos? Usually cultures can't even stay compatible for a few weeks, so doing this is a bit reachy (ex this format will stay until clocks are nanos). Zipkin primarily used epoch micros 

I think the problem of nano started with google's proto extensions which jaeger adopted (not sure why they adopted that, as initially they used zipkin's model which was epoch micros @pavolloffay might know why)

https://github.com/jaegertracing/jaeger-idl/blob/master/proto/api_v2/model.proto#L117-L124

(the adoption maybe was from opencensus who eagerly went to [google proto stuff 4 years ago](https://github.com/census-instrumentation/opencensus-java/pull/50))

Google's proto is epoch seconds and allows an timestamp offset in nanos resolution  https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/timestamp.proto This implies cost as you mention, doing multiplication all the time. Also, it was a data structure so inherently more expensive than a single field. In zi

I believe Jaeger is co-authoring/moving to open-telemetry for data model which uses a single numeric field again:
https://github.com/open-telemetry/opentelemetry-proto/blob/master/opentelemetry/proto/trace/v1/trace.proto#L136-L150

In fact, otel in their own rationale only considered jaeger when thinking about the problem https://github.com/open-telemetry/oteps/blob/600fa63a969c0e1b7ad817469e8dcc518b7561df/text/trace/0059-otlp-trace-data-format.md#trade-offs-and-mitigations

EOF is I think this is just people continually inheriting something from google, but isn't itself relevant for the initial reasons I mentioned.

**Comment by jcchavezs** [comment]
Related note https://github.com/hypertrace/hypertrace/issues/139

**Comment by subintp** [comment]
This issue is creating confusion on the Hypertace UI also.  We are **seeing traces with latency as 0ms** due to this

<img width="1220" alt="Screenshot 2021-04-23 at 11 19 02 AM" src="https://user-images.githubusercontent.com/1742357/115824552-b9083480-a425-11eb-8c8c-6c029d80b9f3.png">
